altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (C) 2017  AltOS-Rust Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! Linked list code for the the memory allocator
+//! This is intended for use by the free_list_allocator
+//!
+
+use core::{mem, ptr};
+
+use alignment;
+
+// BlockHeader keeps track of a free block of memory
+#[cfg(not(test))]
+#[repr(C)]
+pub struct BlockHeader {
+    block_size: usize,
+    next_block: *mut BlockHeader,
+}
+
+impl BlockHeader {
+    const fn new(size: usize) -> Self {
+        BlockHeader {
+            block_size: size,
+            next_block: ptr::null_mut(),
+        }
+    }
+}
+
+// FreeList is a linked list which keeps track of free blocks of memory
+// Free blocks are embedded in the free memory itself
+#[cfg(not(test))]
+pub struct FreeList {
+    block_hdr_size: usize,
+    head: *mut BlockHeader,
+}
+
+// These are (trivially) implemented so FreeList objects can be passed between threads.
+unsafe impl Send for FreeList {}
+unsafe impl Sync for FreeList {}
+
+impl FreeList {
+    pub const fn new() -> Self {
+        FreeList {
+            block_hdr_size: 0,
+            head: ptr::null_mut(),
+        }
+    }
+
+    pub fn init(&mut self, heap_start: usize, heap_size: usize) {
+        self.block_hdr_size = mem::size_of::<BlockHeader>();
+        // Forcing BlockHeader size to be a power of two avoids weird issues with alignment
+        // and extra logic we would have to include elsewhere
+        if !self.block_hdr_size.is_power_of_two() {
+            panic!("block header size is not power of two");
+        }
+        // We adjust the heap starting position and size so that it will initially have
+        // a starting position and size aligned to the block header size.
+        // This potentially leaks a few bytes but it might cause errors if we didn't do it.
+        let use_heap_start = alignment::align_up(heap_start, self.block_hdr_size);
+        let align_diff = use_heap_start - heap_start;
+        // Adjust the heap size down based on alignment change to starting position
+        // and then adjust it down again if it's not aligned to block header size.
+        let use_heap_size = alignment::align_down(heap_size - align_diff, self.block_hdr_size);
```

Comment:
Yes, we might be leaking bytes here. I see what you're getting at with the comments about the alignment functions, but we're doing this for a reason. Basically, we're trying to keep every allocation aligned to the size of BlockHeader.

We run into potential problems if we don't enforce this. Suppose we have a free block with 12 bytes, and an allocation asks for 4 bytes. We have to keep every allocation at least 8 bytes so we have enough space for a new BlockHeader when it gets deallocated, but if we use 8 bytes, we only have 4 left. What do we do with those? We can't create a new free block because it's too small. We could let the allocation use those 4 bytes, but what happens when the program deallocates that memory? It doesn't have any way of knowing that we gave it 12 bytes, so 4 bytes are leaked. The way we avoid these sorts of issues is to force every free block (and consequently, every allocation) to be a multiple of 8, so there should never be a free block with 12 bytes (only 8, 16, etc...). This also means that the starting heap needs to have that property as well.